core/src/main/java/org/apache/spark/shuffle/sort/UnsafeShuffleWriter.java

```diff
@@ -425,14 +425,8 @@ void forceSorterToSpill() throws IOException {
           long bytesToTransfer = partitionLengthInSpill;
           final FileChannel spillInputChannel = spillInputChannels[i];
           final long writeStartTime = System.nanoTime();
-          while (bytesToTransfer > 0) {
-            final long actualBytesTransferred = spillInputChannel.transferTo(
-              spillInputChannelPositions[i],
-              bytesToTransfer,
-              mergedFileOutputChannel);
-            spillInputChannelPositions[i] += actualBytesTransferred;
-            bytesToTransfer -= actualBytesTransferred;
-          }
+          Utils.copyFileStreamNIO(spillInputChannel, mergedFileOutputChannel, bytesToTransfer);
```

Comment:
This is not equivalent to the old code. The `copyFileStreamNIO` method is assuming that you're starting to transfer from position 0 in the input channel, which is only true on the first iteration of the outer loop.

I think you need to add a fourth argument to `copyFileStreamNIO` to specify the starting position of the input channel.